fedora_messaging/twisted/consumer.py

```diff
@@ -49,10 +50,53 @@ def __init__(self, queue=None, callback=None):
         self._channel = None
         # The unique ID for the AMQP consumer.
         self._tag = str(uuid.uuid4())
-        # Used to start and stop the consumer via cancel(), pauseProducing()
+        # Used in the consumer read loop to know when it's being canceled.
         self._running = True
         # The current read loop
         self._read_loop = None
+        # The protocol that currently runs this consumer, used when cancel is
+        # called to remove itself from the protocol and its factory so it doesn't
+        # restart on the next connection.
+        self._protocol = None
 
     def __repr__(self):
         return "Consumer(queue={}, callback={})".format(self.queue, self.callback)
+
+    @defer.inlineCallbacks
+    def cancel(self):
+        """
+        Cancel the consumer and clean up resources associated with it.
+        Consumers that are canceled are allowed to finish processing any
+        messages before halting.
+
+        Returns:
+            defer.Deferred: A deferred that fires when the consumer has finished
+            processing any message it was in the middle of and has been successfully
+            canceled.
+        """
+        # Remove it from protocol and factory so it doesn't restart later.
+        try:
+            del self._protocol._consumers[self.queue]
+        except (KeyError, AttributeError):
+            pass
+        try:
+            del self._protocol.factory._consumers[self.queue]
+        except (KeyError, AttributeError):
+            pass
+        # Signal to the _read loop it's time to stop and wait for it to finish
+        # with whatever message it might be working on, then wait for the deferred
+        # to fire which indicates it is done.
+        self._running = False
+        yield self._read_loop
+        try:
+            yield self._channel.basic_cancel(consumer_tag=self._tag)
+        except pika.exceptions.AMQPChannelError:
+            # Consumers are tied to channels, so if this channel is dead the
+            # consumer should already be canceled (and we can't get to it anyway)
+            pass
+        try:
+            yield self._channel.close()
```

Comment:
There can be, but we don't do that because if any of the other consumers do something that results in an exception, pika (or maybe it's the broker?) kills the channel. We used to use one channel for all the consumers and then they all died if you accidentally tried to declare a queue you didn't have permission to, for example.